src/scheduler/gc_work.rs

```diff
@@ -683,40 +709,73 @@ impl<VM: VMBinding> ProcessEdgesWork for SFTProcessEdges<VM> {
     }
 
     fn create_scan_work(&self, nodes: Vec<ObjectReference>, roots: bool) -> ScanObjects<Self> {
-        ScanObjects::<Self>::new(nodes, false, roots)
+        ScanObjects::<Self>::new(nodes, false, roots, self.bucket)
     }
 }
-
-struct ProcessEdgesWorkRootsWorkFactory<E: ProcessEdgesWork> {
-    mmtk: &'static MMTK<E::VM>,
+struct ProcessEdgesWorkRootsWorkFactory<
+    VM: VMBinding,
+    E: ProcessEdgesWork<VM = VM>,
+    I: ProcessEdgesWork<VM = VM>,
+> {
+    mmtk: &'static MMTK<VM>,
+    phantom: PhantomData<(E, I)>,
 }
 
-impl<E: ProcessEdgesWork> Clone for ProcessEdgesWorkRootsWorkFactory<E> {
+impl<VM: VMBinding, E: ProcessEdgesWork<VM = VM>, I: ProcessEdgesWork<VM = VM>> Clone
+    for ProcessEdgesWorkRootsWorkFactory<VM, E, I>
+{
     fn clone(&self) -> Self {
-        Self { mmtk: self.mmtk }
+        Self {
+            mmtk: self.mmtk,
+            phantom: PhantomData,
+        }
     }
 }
 
-impl<E: ProcessEdgesWork> RootsWorkFactory<EdgeOf<E>> for ProcessEdgesWorkRootsWorkFactory<E> {
+impl<VM: VMBinding, E: ProcessEdgesWork<VM = VM>, I: ProcessEdgesWork<VM = VM>>
+    RootsWorkFactory<EdgeOf<E>> for ProcessEdgesWorkRootsWorkFactory<VM, E, I>
+{
     fn create_process_edge_roots_work(&mut self, edges: Vec<EdgeOf<E>>) {
         crate::memory_manager::add_work_packet(
             self.mmtk,
             WorkBucketStage::Closure,
-            E::new(edges, true, self.mmtk),
+            E::new(edges, true, self.mmtk, WorkBucketStage::Closure),
         );
     }
 
     fn create_process_node_roots_work(&mut self, nodes: Vec<ObjectReference>) {
-        // We want to use E::create_scan_work.
-        let process_edges_work = E::new(vec![], true, self.mmtk);
-        let work = process_edges_work.create_scan_work(nodes, true);
-        crate::memory_manager::add_work_packet(self.mmtk, WorkBucketStage::Closure, work);
+        crate::memory_manager::add_work_packet(
+            self.mmtk,
+            WorkBucketStage::Closure,
+            ProcessRootNode::<VM, I, E>::new(nodes, WorkBucketStage::Closure),
+        );
```

Comment:
The `create_process_node_roots_work` is supposed to list "black roots" which pin the objects in `nodes` but not their children.  The objects in `nodes` must not move.  However, if you put the `ProcessRootNode` work packet in `Closure`, other work packets that can move objects may be executed before this `ProcessRootNode` work packet.  When this `ProcessRootNode` work packet is executed, it may find some objects in `nodes` are already moved.

The right way is putting `ProcessRootNode` in a new `NTPinTrace` bucket (see comments in `work_bucket.rs`), but putting its derived `ScanObjects` and `ProcessEdgesWork` work packets into the regular `Closure` bucket.  In this way, it will guarantee that objects in `nodes` are visited before other non-pinning edges visit those objects.

```suggestion
        crate::memory_manager::add_work_packet(
            self.mmtk,
            WorkBucketStage::NTPinTrace,
            ProcessRootNode::<VM, I, E>::new(nodes, WorkBucketStage::Closure),
        );
```

Reply:
For normal root nodes, we can put the `ProcessRootNode` work to the `TPinClosure`, but allow the packets generated from the work to  be put to the normal closure bucket. In this case, we do not need a third closure bucket other than TPin closure and normal closure.

Reply:
> For normal root nodes, we can put the `ProcessRootNode` work to the `TPinClosure`, but allow the packets generated from the work to be put to the normal closure bucket. In this case, we do not need a third closure bucket other than TPin closure and normal closure.

That won't work, for the same reason why black roots cannot be put into TPinClosure.  If an object is traced from a normal root before it is traced from a TPin (red) root, its children will be put into `Closure`.  When it is then traced from a TPin (red) root, it will find it already marked, and will not enqueue its children in `TPinClosure`.  As a result, its children will not be pinned, but they should be pinned because they are reachable from a TPin (red) root.

Reply:
I see.

Reply:
I've created a `NodeRootsTrace` to try to fix the issue (see the updated code). I called it NodeRoots because it is really specific to node roots, in which the root node shouldn't move, but there is no restriction about their transitive closure.